classes/PodsAPI.php

```diff
@@ -2885,42 +2885,106 @@ public function save_field( $params, $table_operation = true, $sanitized = false
+				if ( ( $field['type'] !== $old_type || $old_simple !== $simple ) && empty( $definition ) ) {
+					$definition_mode = 'drop';
 				} elseif ( 0 < strlen( $definition ) ) {
 					if ( $old_name !== $field['name'] || $old_simple !== $simple || $old_definition !== $definition ) {
-						$test = false;
+						$definition_mode = 'add';
 
 						if ( 0 < strlen( $old_definition ) ) {
-							$test = pods_query( "ALTER TABLE `@wp_pods_{$params->pod}` CHANGE `{$old_name}` {$definition}", false );
-						}
-
-						// If the old field doesn't exist, continue to add a new field
-						if ( false === $test ) {
-							pods_query( "ALTER TABLE `@wp_pods_{$params->pod}` ADD COLUMN {$definition}", __( 'Cannot create new field', 'pods' ) );
+							$definition_mode = 'change';
 						}
 					} elseif ( null !== $old_definition && $definition !== $old_definition ) {
-						$test = pods_query( "ALTER TABLE `@wp_pods_{$params->pod}` CHANGE `{$old_name}` {$definition}", false );
-
-						// If the old field doesn't exist, continue to add a new field
-						if ( false === $test ) {
-							pods_query( "ALTER TABLE `@wp_pods_{$params->pod}` ADD COLUMN {$definition}", __( 'Cannot create new field', 'pods' ) );
-						}
+						$definition_mode = 'change';
 					}
 				}
 			} elseif ( 0 < strlen( $definition ) ) {
-				$test = false;
+				$definition_mode = 'add';
 
 				if ( 0 < strlen( $old_definition ) ) {
-					$test = pods_query( "ALTER TABLE `@wp_pods_{$params->pod}` CHANGE `" . $field['name'] . "` {$definition}", false );
+					$definition_mode = 'change';
+				}
+			}
+
+			if ( 'bypass' !== $definition_mode ) {
+				/**
+				 * Allow hooking into nefore the table has been altered for any custom functionality needed.
+				 *
+				 * @since TBD
+				 *
+				 * @param string             $definition_mode The definition mode used for the table field.
+				 * @param Pods\Whatsit\Pod   $pod             The pod object.
+				 * @param string             $type            The field type.
+				 * @param Pods\Whatsit\Field $field           The field object.
+				 * @param array              $extra_info      {
+				 *      Extra information about the field.
+				 *
+				 *      @type bool        $simple Whether the field is a simple tableless field.
+				 *      @type string      $definition The field definition.
+				 *      @type null|string $old_name The old field name (if preexisting).
+				 *      @type null|string $old_definition The old field definition (if preexisting).
+				 *      @type null|array  $old_options The old field options (if preexisting).
+				 * }
+				 */
+				do_action( 'pods_api_save_field_table_pre_alter', $definition_mode, $pod, $field['type'], $field, [
+					'simple'         => $simple,
+					'definition'     => $definition,
+					'old_name'       => $old_name,
+					'old_definition' => $old_definition,
+					'old_options'    => $old_options,
+				] );
```

Comment:
:no_entry_sign: Closing parenthesis of a multi-line function call must be on a line by itself

<sub>hash: b6545c5257e0b7d1f74a80e0355a22f9</sub>